android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java

```diff
@@ -93,8 +96,33 @@
 	private TiWeakList<OnPrepareOptionsMenuEvent> onPrepareOptionsMenuListeners = new TiWeakList<OnPrepareOptionsMenuEvent>();
 	private APSAnalytics analytics = APSAnalytics.getInstance();
 
-	public static KrollObject storageCallbackContext, cameraCallbackContext, contactsCallbackContext, oldCalendarCallbackContext, calendarCallbackContext, locationCallbackContext;
-	public static KrollFunction storagePermissionCallback, cameraPermissionCallback, contactsPermissionCallback, oldCalendarPermissionCallback, calendarPermissionCallback, locationPermissionCallback;
+
+	public static class PermissionContextData {
+		private final Integer requestCode;
+		private final KrollObject context;
+		private final KrollFunction callback;
+
+		public PermissionContextData(Integer requestCode, KrollFunction callback,
+									 KrollObject context) {
+			this.requestCode = requestCode;
+			this.callback = callback;
+			this.context = context;
+		}
+
+		public Integer getRequestCode() {
+			return requestCode;
+		}
+
+		public KrollFunction getCallback() {
+			return callback;
+		}
+
+		public KrollObject getContext() {
+			return context;
+		}
+	}
+
+	private static ConcurrentHashMap<Integer,PermissionContextData> callbackDataByPermission = new ConcurrentHashMap<Integer, PermissionContextData>();
```

Comment:
Would this cause any memory leaks on Activities (Windows)? Might need to check when using this, when the Window is closed, does the Activity get released by doing Heap dumps.